File: src/test/java/com/syfe/personalfinance/service/UserServiceTest.java
package com.syfe.personalfinance.service;

import com.syfe.personalfinance.dto.AuthDto;
import com.syfe.personalfinance.dto.AuthDto.UserResponse;
import com.syfe.personalfinance.entity.User;
import com.syfe.personalfinance.exception.ConflictException;
import com.syfe.personalfinance.mapper.UserMapper;
import com.syfe.personalfinance.repository.UserRepository;
import com.syfe.personalfinance.service.impl.UserServiceImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class UserServiceTest {

    @Mock
    private UserRepository userRepository;

    @Mock
    private UserMapper userMapper;

    @Mock
    private PasswordEncoder passwordEncoder;

    private UserService userService;

    @BeforeEach
    void setUp() {
        userService = new UserServiceImpl(userRepository, userMapper, passwordEncoder);
    }

    @Test
    void register_Success() {
        AuthDto.RegisterRequest request = AuthDto.RegisterRequest.builder()
                .username("test@syfe.com")
                .password("SecurePassword123!")
                .fullName("Test User")
                .build();

        User userEntity = User.builder()
                .username("test@syfe.com")
                .password("encodedPassword")
                .fullName("Test User")
                .build();

        UserResponse expectedResponse = UserResponse.builder()
                .id(1L)
                .username("test@syfe.com")
                .fullName("Test User")
                .build();

        when(userRepository.existsByUsername(request.getUsername())).thenReturn(false);
        when(userMapper.toEntity(request)).thenReturn(userEntity);
        when(passwordEncoder.encode(request.getPassword())).thenReturn("encodedPassword");
        when(userRepository.save(any(User.class))).thenReturn(userEntity);
        when(userMapper.toResponse(any(User.class))).thenReturn(expectedResponse);

        UserResponse actualResponse = userService.register(request);

        assertNotNull(actualResponse);
        assertEquals("test@syfe.com", actualResponse.getUsername());
        verify(userRepository, times(1)).save(any(User.class));
    }

    @Test
    void register_ConflictingUsername_ThrowsException() {
        AuthDto.RegisterRequest request = AuthDto.RegisterRequest.builder()
                .username("duplicate@syfe.com")
                .password("SecurePassword123!")
                .fullName("Duplicate User")
                .build();

        when(userRepository.existsByUsername(request.getUsername())).thenReturn(true);

        assertThrows(ConflictException.class, () -> userService.register(request));
        verify(userRepository, never()).save(any(User.class));
    }

    @Test
    void loadUserByUsername_Success() {
        User userEntity = User.builder()
                .username("user@syfe.com")
                .password("encoded_pass")
                .build();

        when(userRepository.findByUsername("user@syfe.com")).thenReturn(Optional.of(userEntity));

        UserDetails userDetails = userService.loadUserByUsername("user@syfe.com");

        assertNotNull(userDetails);
        assertEquals("user@syfe.com", userDetails.getUsername());
        assertEquals("encoded_pass", userDetails.getPassword());
    }

    @Test
    void loadUserByUsername_NotFound_ThrowsException() {
        when(userRepository.findByUsername("nonexistent@syfe.com")).thenReturn(Optional.empty());

        assertThrows(UsernameNotFoundException.class, () -> userService.loadUserByUsername("nonexistent@syfe.com"));
    }
}

File: src/test/java/com/syfe/personalfinance/service/UserServiceTest.kt
package com.syfe.personalfinance.service

import com.syfe.personalfinance.dto.AuthDto
import com.syfe.personalfinance.dto.AuthDto.UserResponse
import com.syfe.personalfinance.entity.User
import com.syfe.personalfinance.exception.ConflictException
import com.syfe.personalfinance.mapper.UserMapper
import com.syfe.personalfinance.repository.UserRepository
import com.syfe.personalfinance.service.impl.UserServiceImpl
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class UserServiceTest {

    @Mock
    private lateinit var userRepository: UserRepository

    @Mock
    private lateinit var userMapper: UserMapper

    @Mock
    private lateinit var passwordEncoder: PasswordEncoder

    private lateinit var userService: UserService

    @BeforeEach
    fun setUp() {
        userService = UserServiceImpl(userRepository, userMapper, passwordEncoder)
    }

    @Test
    fun register_Success() {
        val request = AuthDto.RegisterRequest(
            username = "test@syfe.com",
            password = "SecurePassword123!",
            fullName = "Test User"
        )

        val userEntity = User(
            username = "test@syfe.com",
            password = "encodedPassword",
            fullName = "Test User"
        )

        val expectedResponse = UserResponse(
            id = 1L,
            username = "test@syfe.com",
            fullName = "Test User"
        )

        whenever(userRepository.existsByUsername(request.username)).thenReturn(false)
        whenever(userMapper.toEntity(request)).thenReturn(userEntity)
        whenever(passwordEncoder.encode(request.password)).thenReturn("encodedPassword")
        whenever(userRepository.save(any<User>())).thenReturn(userEntity)
        whenever(userMapper.toResponse(any<User>())).thenReturn(expectedResponse)

        val actualResponse = userService.register(request)

        assertNotNull(actualResponse)
        assertEquals("test@syfe.com", actualResponse.username)
        verify(userRepository, times(1)).save(any<User>())
    }

    @Test
    fun register_ConflictingUsername_ThrowsException() {
        val request = AuthDto.RegisterRequest(
            username = "duplicate@syfe.com",
            password = "SecurePassword123!",
            fullName = "Duplicate User"
        )

        whenever(userRepository.existsByUsername(request.username)).thenReturn(true)

        assertThrows(ConflictException::class.java) { userService.register(request) }
        verify(userRepository, never()).save(any<User>())
    }

    @Test
    fun loadUserByUsername_Success() {
        val userEntity = User(
            username = "user@syfe.com",
            password = "encoded_pass"
        )

        whenever(userRepository.findByUsername("user@syfe.com")).thenReturn(Optional.of(userEntity))

        val userDetails = userService.loadUserByUsername("user@syfe.com")

        assertNotNull(userDetails)
        assertEquals("user@syfe.com", userDetails.username)
        assertEquals("encoded_pass", userDetails.password)
    }

    @Test
    fun loadUserByUsername_NotFound_ThrowsException() {
        whenever(userRepository.findByUsername("nonexistent@syfe.com")).thenReturn(Optional.empty())

        assertThrows(UsernameNotFoundException::class.java) { userService.loadUserByUsername("nonexistent@syfe.com") }
    }
}
